lib/elixir/lib/calendar/iso.ex

```diff
@@ -663,6 +664,147 @@ defmodule Calendar.ISO do
     end
   end
 
+  @doc """
+  Parses an ISO 8601 formatted duration string to a list of `Duration` compabitble unit pairs.
+
+  """
+  @doc since: "1.17.0"
+  @spec parse_duration(String.t()) :: {:ok, [Duration.unit_pair()]} | {:error, atom()}
+  def parse_duration("P" <> string) do
```

Comment:
An idea to shorten the code for `parse_duration/1`. The following code makes all `parse_date_duration/3` and `parse_time_duration/3` functions obsolete.
```elixir
  def parse_duration("P" <> string), do: parse_duration(string, :date, %{}, "")
  def parse_duration("T" <> string), do: parse_duration(string, :time, %{}, "")
  def parse_duration(_), do: {:error, :invalid_duration}

  duration_parser = %{
    date: [year: "Y", month: "M", day: "D", week: "W"],
    time: [hour: "H", minute: "M", second: "S"]
  }

  defp parse_duration(_, _part, {:error, error}, _), do: {:error, error}

  for {part, units} <- duration_parser,
      {unit, designator} <- units do
    defp parse_duration(<<unquote(designator), rest::binary>>, unquote(part), duration, buffer) do
      duration = parse_buffer_duration(unquote(unit), duration, buffer)
      parse_duration(rest, unquote(part), duration, "")
    end
  end

  defp parse_duration(<<c, rest::binary>>, part, duration, buffer)
       when c in ?0..?9 or c in [?., ?-] do
    parse_duration(rest, part, duration, <<buffer::binary, c>>)
  end

  defp parse_duration(<<"T", rest::binary>>, :date, duration, _buffer) do
    parse_duration(rest, :time, duration, "")
  end

  defp parse_duration(<<>>, _part, duration, ""), do: {:ok, Keyword.new(duration)}

  defp parse_duration(_, _, _, _), do: {:error, :invalid_character}

```